File: src/main/java/mods/eln/misc/BoundingBox.java
package mods.eln.misc;

import com.google.common.base.Objects;
import net.minecraft.util.math.Vec3d;


public class BoundingBox {
    public final Vec3d min, max;

    public BoundingBox(float xMin, float xMax, float yMin, float yMax, float zMin, float zMax) {
        min = new Vec3d(xMin, yMin, zMin);
        max = new Vec3d(xMax, yMax, zMax);
    }

    public static BoundingBox mergeIdentity() {
        return new BoundingBox(
            Float.POSITIVE_INFINITY,
            Float.NEGATIVE_INFINITY,
            Float.POSITIVE_INFINITY,
            Float.NEGATIVE_INFINITY,
            Float.POSITIVE_INFINITY,
            Float.NEGATIVE_INFINITY
        );
    }

    public BoundingBox merge(BoundingBox other) {
        return new BoundingBox(
            (float) Math.min(min.xCoord, other.min.xCoord),
            (float) Math.max(max.xCoord, other.max.xCoord),
            (float) Math.min(min.yCoord, other.min.yCoord),
            (float) Math.max(max.yCoord, other.max.yCoord),
            (float) Math.min(min.zCoord, other.min.zCoord),
            (float) Math.max(max.zCoord, other.max.zCoord)
        );
    }

    public Vec3d centre() {
        return new Vec3d(
            min.xCoord + (max.xCoord - min.xCoord) / 2,
            min.yCoord + (max.yCoord - min.yCoord) / 2,
            min.zCoord + (max.zCoord - min.zCoord) / 2
        );
    }

    @Override
    public String toString() {
        return Objects.toStringHelper(this)
            .add("min", min)
            .add("max", max)
            .toString();
    }
}

File: src/main/java/mods/eln/misc/BoundingBox.kt
package mods.eln.misc

import net.minecraft.util.math.Vec3d

@Deprecated("The Minecraft one should do fine.")
class BoundingBox(xMin: Float, xMax: Float, yMin: Float, yMax: Float, zMin: Float, zMax: Float) {
    val min: Vec3d = Vec3d(xMin.toDouble(), yMin.toDouble(), zMin.toDouble())
    val max: Vec3d = Vec3d(xMax.toDouble(), yMax.toDouble(), zMax.toDouble())

    fun merge(other: BoundingBox): BoundingBox {
        return BoundingBox(
                Math.min(min.x, other.min.x).toFloat(),
                Math.max(max.x, other.max.x).toFloat(),
                Math.min(min.y, other.min.y).toFloat(),
                Math.max(max.y, other.max.y).toFloat(),
                Math.min(min.z, other.min.z).toFloat(),
                Math.max(max.z, other.max.z).toFloat()
        )
    }

    fun centre(): Vec3d {
        return Vec3d(
                min.x + (max.x - min.x) / 2,
                min.y + (max.y - min.y) / 2,
                min.z + (max.z - min.z) / 2
        )
    }

    companion object {
        fun mergeIdentity(): BoundingBox {
            return BoundingBox(
                    java.lang.Float.POSITIVE_INFINITY,
                    java.lang.Float.NEGATIVE_INFINITY,
                    java.lang.Float.POSITIVE_INFINITY,
                    java.lang.Float.NEGATIVE_INFINITY,
                    java.lang.Float.POSITIVE_INFINITY,
                    java.lang.Float.NEGATIVE_INFINITY
            )
        }
    }

    override fun toString() = "min: $min, max: $max"
}
